### ISSUE TITLE
Automatically collapse the go-plus panel when saving and no errors are present

### ISSUE THREAD

**Issue by Asday** [joefitzgerald/go-plus#606]
When I save a go file, and I've missed a bracket or mismatched a type, go-plus jumps open and tells me, taking up a large amount of space on my screen.  This is a good thing and I like it.

I dislike that the panel doesn't go away when I fix the errors and save the file, instead, it goes mostly blank and shows me the name of the package I'm working on.

I looked through the settings for such an option, and was unable to find one.  May we have one, please?

**Comment by lloiser** [comment]
Please answer the questions instead of deleting the text when creating a new issue...

The panel you are seeing is actually not the go-plus panel but from the [`linter`](https://github.com/steelbrain/linter) package. 
This linter package had a major update recently that might be the cause for this problem. You can try "downgrading" the linter package using `apm install linter@1.11.23` and see if the problem still occurs.

Please post the versions of `go-plus` and `linter` you are currently using (either via the settings panel or `apm list`)

**Comment by Asday** [comment]
>Please answer the questions instead of deleting the text when creating a new issue...

My apologies.  I read through them and misapprehended that they were not useful information for a feature request.

Downgrading to `1.11.23` was a good idea, but didn't fix the annoyance.  I haven't used go-plus enough in the past to say for sure whether it ever didn't do this.

I am on linter 2.1.2, and go-plus 5.3.4.

**Comment by joefitzgerald** [comment]
If I am interpreting the request correctly, the desired behavior is:

* go-plus panel is hidden
* lots of great work is done, ideally by writing tests first
* a test fails, and the go-plus panel pops up and tells you why
* you write code to fix the test, and the panel is hidden again
* rinse and repeat

Is that roughly accurate?

**Comment by Asday** [comment]
Correct, except being new to this, I'm not so much failing tests as failing compiles; getting errors mentioning incorrect argument types, missing brackets, and so forth.

>lots of great work is done

Haha :D

**Comment by zmb3** [comment]
This all goes back to differentiating test failure and build failure. Which
we can do when we better orchestrate the save event.

Ideally a failed build won't be interpreted as a test failure and the test
panel wouldn't come up in this case. You would instead be notified by a
linter or build panel.

**Comment by zer09** [comment]
I just want to  ask if there is a corresponding key-binding to hide the go-plus panel? Because I would like to hide it if the error of the code is fixed.

BTW I am new in atom.io

Thanks

**Comment by joefitzgerald** [comment]
Yes, they escape key willl hide it.

**Comment by zer09** [comment]
@joefitzgerald thank you, but it will just work if the panel is on focus. If I will focus on the editor then pressing the escape key, the panel won't hide.

**Comment by joefitzgerald** [comment]
You want the `toggle-panel` command: https://github.com/joefitzgerald/go-plus/blob/master/keymaps/go-plus.json#L3

**Comment by zer09** [comment]
@joefitzgerald Thank you so much.

**Comment by Zamiell** [comment]
I'm (still) having the same problem as the original poster.

> lloiser commented on Mar 23
> The panel you are seeing is actually not the go-plus panel but from the linter package.

Is that true? For example, look at the follow screenshot of my Atom window:

![go-plus](https://user-images.githubusercontent.com/5511220/29186922-f0bd3c32-7ddb-11e7-8ef2-db4f4af3129a.png)

It is my understanding that the **red** square is provided by the `linter` package, and the **green** square is provided by the `go-plus` package.

The red square does goes away automatically when there are 0 errors, presumably because I have the following checkbox checked in the `linter-ui-default` settings menu:

![image](https://user-images.githubusercontent.com/5511220/29186988-2c56583c-7ddc-11e7-937f-25d84d4f7c33.png)

This is all well and great. However, the green square will never go away, even if there are 0 errors, reporting this perpetually:
```
Running go install .
```

I looked through the settings menu for `go-plus`, and I didn't see a similar "Hide Panel When No Issues" checkbox. I apologize for replying to a closed issue, but I seem to be rather confused about this feature!

**Comment by zmb3** [comment]
Your understanding is correct.

You just need to collapse the go-plus panel (in green) once (either by pressing <kbd>esc</kbd>, or clicking the `X`), and then it won't come up again on save unless there is an error.

If you don't want it to come up even when there is an error, you can disable the "focus the go-plus panel on failure" setting.

If you don't want to have to collapse it manually the first time, you can set "panel display mode" to _closed_.

tldr; the panel is never automatically hidden.  You can hide it or tell it to be hidden by default, and you can prevent it from showing itself when there are errors.

**Comment by Zamiell** [comment]
@zmb3 Looks like the information in this thread could use an update, as `go-plus` has been updated and the circumstances have changed, and this URL is still the number one google result for "go-plus hide panel on success", and similar searches.

By default, on a fresh Atom install and a fresh `go-plus` install, the `linter` package is not used whatsoever, so for future readers, all of the above information about `linter` can probably be ignored.

Now, `go-plus` will just use the default go-plus panel. However, it is unclear how to solve the problem that the original poster presents:

> When I save a go file, and I've missed a bracket or mismatched a type, go-plus jumps open and tells me, taking up a large amount of space on my screen. This is a good thing and I like it. I dislike that the panel doesn't go away when I fix the errors and save the file, instead, it goes mostly blank and shows me the name of the package I'm working on.

Is this currently possible with the new panel, or is this a regression in `go-plus`?

**Comment by zmb3** [comment]
I’m not sure what you mean by regression, because that implies something that worked before no longer works now. 

The behavior of the panel has not changed with the new updates. There is still an option to bring it up automatically when a failure occurs. There is not a feature to automatically hide the panel when errors go away, mostly because it is not an errors panel. It has lots of useful info that’s relevant whether or not you have an error in your code.  I personally prefer to leave it open so that I can see the test output on every save. 

Additionally, since the panel is part of an Atom dock now, collapsing the panel means potentially also hiding any other panels that are sharing that dock. That’s not something I think we should do. 

There is a key binding to show/hide the panel. It’s a bit long but you can change the key map to any key combination that works well for you. Does that help the workflow in any way?

**Comment by Zamiell** [comment]
I appreciate the clarification; thank you. Unfortunately, having to manually hit a hotkey after every single linter fix is fairly arduous and does not really help the workflow.

I think it is a regression because the "feature" of having the panel disappear was something that did work before and no longer works now, although I freely admit I am most likely confused about `go-plus` requiring the `linter` package earlier in its lifespan and the documented interaction between the two.

Nonetheless, as a note to future readers, know that I was able to restore the functionality by re-installing the `linter` package. After that, I got a popup about having `linter` conflict with some diagnostics package, so I chose to use `linter` over that one. Then, I checked the box in the `linter-ui-default` settings that said "Show Panel", and then I hit the X on the `go-plus` panel. And now it shows on errors and hides on success. Hooray!
